ts/sdk/src/jitter.ts

```diff
@@ -136,57 +188,256 @@ export class Jitter {
+		const params = isVariant(order.marketType, 'perp')
+			? this.perpParams.get(order.marketIndex)
+			: this.spotParams.get(order.marketIndex);
+		const oraclePrice = isVariant(order.marketType, 'perp')
+			? this.driftClient.getOracleDataForPerpMarket(order.marketIndex)
+			: this.driftClient.getOracleDataForSpotMarket(order.marketIndex);
+
+		const makerOrderDir = isVariant(order.direction, 'long') ? 'sell' : 'buy';
+		const auctionStartPrice = convertToNumber(
+			isVariant(order.orderType, 'oracle')
+				? getAuctionPriceForOracleOffsetAuction(
+						order,
+						order.slot.toNumber(),
+						oraclePrice.price
+				  )
+				: order.auctionStartPrice,
+			PRICE_PRECISION
+		);
+		const auctionEndPrice = convertToNumber(
+			isVariant(order.orderType, 'oracle')
+				? getAuctionPriceForOracleOffsetAuction(
+						order,
+						order.slot.toNumber() + order.auctionDuration - 1,
+						oraclePrice.price
+				  )
+				: order.auctionEndPrice,
+			PRICE_PRECISION
+		);
+
+		const bid = isVariant(params.priceType, 'oracle')
+			? convertToNumber(oraclePrice.price.sub(params.bid), PRICE_PRECISION)
+			: convertToNumber(params.bid, PRICE_PRECISION);
+		const ask = isVariant(params.priceType, 'oracle')
+			? convertToNumber(oraclePrice.price.add(params.ask), PRICE_PRECISION)
+			: convertToNumber(params.ask, PRICE_PRECISION);
+
+		let slotsTilCross = 0;
+		let willCross = false;
+		const stepSize =
+			(auctionEndPrice - auctionStartPrice) / (order.auctionDuration - 1);
+		while (slotsTilCross < order.auctionDuration) {
+			if (makerOrderDir === 'buy') {
+				if (
+					convertToNumber(
+						getAuctionPrice(
+							order,
+							order.slot.toNumber() + slotsTilCross,
+							oraclePrice.price
+						),
+						PRICE_PRECISION
+					) <= bid
+				) {
+					willCross = true;
+					break;
+				}
+			} else {
+				if (
+					convertToNumber(
+						getAuctionPrice(
+							order,
+							order.slot.toNumber() + slotsTilCross,
+							oraclePrice.price
+						),
+						PRICE_PRECISION
+					) >= ask
+				) {
+					willCross = true;
+					break;
 				}
-				i++;
 			}
+			slotsTilCross++;
+		}
 
-			this.onGoingAuctions.delete(orderSignature);
+		return {
+			slotsTilCross,
+			willCross,
+			bid,
+			ask,
+			auctionStartPrice,
+			auctionEndPrice,
+			stepSize,
+			oraclePrice,
 		};
 	}
 
+	async waitForSlotOrCrossOrExpiry(
+		targetSlot: number,
+		order: Order,
+		initialDetails: AuctionAndOrderDetails
+	): Promise<{ slot: number; updatedDetails: AuctionAndOrderDetails }> {
+		const auctionEndSlot = order.auctionDuration + order.slot.toNumber();
+		let currentDetails: AuctionAndOrderDetails = initialDetails;
+		let willCross = initialDetails.willCross;
+		return new Promise((resolve) => {
+			// Immediately return if we are past target slot
+
+			// Otherwise listen for new slots in case we hit the target slot and we're gonna cross
+			this.slotSubscriber.eventEmitter.on('newSlot', (slot) => {
```

Comment:
do you need to delete these callbacks after the auction expires? over time it'll get slower and slower if there are a bunch of callbacks?

Reply:
or mybe it should only be on callback?

Reply:
fixed this